nni/mutable/utils.py

```diff
@@ -0,0 +1,300 @@
+        """
+        if self.path is None:
+            raise ValueError('label_scope is not entered yet.')
+        return '/'.join(self.path)
+
+    def __repr__(self):
+        return f'label_scope({self.absolute_scope!r})'
+
+    def next_label(self) -> str:
+        """Generate the "name" part"""
+        return str(uid(self.absolute_scope))
+
+    @staticmethod
+    def current() -> label_scope | None:
+        """Fetch the nearest label scope activated by ``with``.
+
+        If label scope is never used, or we are currently within no with-block,
+        return none.
+
+        Examples
+        --------
+        >>> with label_scope() as scope1:
+        ...     # somewhere in the middle of the code.
+        ...     label_scope.current()     # Return scope1
+        """
+        try:
+            return ContextStack.top(_LABEL_NAMESPACE_CONTEXT_KEY)
+        except NoContextError:
+            return None
+
+    @staticmethod
+    def global_() -> label_scope:
+        """Fetch the global label scope.
+
+        This label scope can be created on-the-fly and can live without the with-blocks.
+        """
+        return label_scope('global', _path=['global'])
+
+
+def auto_label(label: str | None = None, scope: label_scope | None = None) -> str:
+    """Automatically generate a formatted and reproducible label.
+
+    In case ``label`` is not set, the label name will use the uid sequence of ``scope``.
+    If ``scope`` is not set, it will fetch the nearest scope.
+    If no scope is found, it will use the global scope (:meth:`label_scope.global_`).
+
+    If the scope is found and is not global scope, the scope's full name will be prepended to the label,
+    i.e., the label will then be formatted as ``{scope}/{label}``.
+    Otherwise, there are two cases. Firstly, if label is manually specified, it will be returned directly.
+    Secondly, we rely on the global scope to generate the label name, and the scope name will still be prepended.
+    The rules can be better explained with the examples below.
+
+    Notes
+    -----
+    We always recommend specifying the label manually.
+
+    Parameters
+    ----------
+    label
+        The label name to use. If not specified, it will be generated by the scope.
+    scope
+        The scope to use. If not specified, the nearest scope will be used.
+
+    Examples
+    --------
+    >>> label1 = auto_label('bar')          # bar, because the scope is global
+    >>> label2 = auto_label()               # global/1, because label is not provided
+    >>> with label_scope('foo'):
+    ...     label3 = auto_label()           # foo/1, because in the scope "foo"
+    >>> with label_scope():
+    ...     label4 = auto_label()           # param/1, default key is used
+    >>> with label_scope('another'):
+    ...     label5 = auto_label()           # another/1
+    ...     label6 = auto_label('thing')    # another/thing
+    ...     label7 = auto_label()           # another/2
```

Comment:
what if the following:
```
with label_scope('test'):
    label1 = auto_label('hello', 'world') #??
```

Reply:
This is not valid. It must be

```python
label1 = auto_label('hello', label_scope('world'))
```

I think it's too advanced for regular users. I added a test to make sure it complains "ValueError: label_scope is not entered yet."